plugin/cog/cog-commons/src/main/java/it/geosolutions/imageioimpl/plugins/cog/RangeReader.java

```diff
@@ -0,0 +1,109 @@
+ *    either version 3 of the License, or (at your option) any later version.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package it.geosolutions.imageioimpl.plugins.cog;
+
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.logging.Logger;
+
+/**
+ * @author joshfix
+ * Created on 2019-08-21
+ */
+public abstract class RangeReader {
+
+    protected URI uri;
+    protected ByteBuffer buffer;
+    protected int filesize = -1;
+    protected int headerLength = 16384;
+
+    private final static Logger LOGGER = Logger.getLogger(RangeReader.class.getName());
+
+    public RangeReader(URI uri) {
+        this.uri = uri;
+    }
+
+    /**
+     * Prevents making new range requests for image data that overlap with the header range that has already been read
+     *
+     * @param ranges
+     * @return
+     */
+    protected long[][] reconcileRanges(long[][] ranges) {
+        boolean modified = false;
+        List<long[]> newRanges = new ArrayList<>();
+        for (int i = 0; i < ranges.length; i++) {
+            if (ranges[i][0] < headerLength - 1) {
+                // this range starts inside of what we already read for the header
+                modified = true;
+                if (ranges[i][1] <= headerLength - 1) {
+                    // this range is fully inside the header which was already read; discard this range
+                    LOGGER.fine("Removed range " + ranges[i][0] + "-" + ranges[i][1] + " as it lies fully within"
+                            + " the data already read in the header request");
+                } else {
+                    // this range starts inside the header range, but ends outside of it.
+                    // add a new range that starts at the end of the header range
+                    long[] newRange = new long[]{headerLength - 1, ranges[i][1]};
+
+                    newRanges.add(newRange);
+                    LOGGER.fine("Modified range " + ranges[i][0] + "-" + ranges[i][1]
+                            + " to " + headerLength + "-" + ranges[i][1] + " as it overlaps with data previously"
+                            + " read in the header request");
+                }
+            } else {
+                // fully outside the header area, keep the range
+                newRanges.add(ranges[i]);
+            }
+        }
+
+        if (modified) {
+            return newRanges.toArray(new long[][]{});
+        } else {
+            LOGGER.fine("No ranges modified.");
+            return ranges;
+        }
+    }
+
+    public void setFilesize(int filesize) {
+        buffer = ByteBuffer.allocate(filesize);
```

Comment:
Wow, danger zone! The byte buffer here is actually allocated in memory, so if one has to deal with a bigtiff that's say 100GB in the cloud, the code will actually allocate in the local memory 100GB, even if the current read only aims at reading one tile?
See: https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#allocate(int)

If I understand correctly, the current implementation works for (very) small images and light load, but it's deadly easy to make it go OOM. It would need some sort of sparse byte buffer, or, instead of delegating to a MemoryCacheImageInputStream, build its own implementation that only keeps in memory the tiles needed for the current read... should be doable by keeping an "index" of offsets pointing at the actual bytes, and doing some offset math whenever a read is called.

Reply:
Thanks for pointing this out.  In fact I did not take into account very large files or consider the ramifications of blindly allocating the full filesize in the byte buffer.  I'm going to spend a bit of time contemplating the best way to handle this.  My initial design was to keep the `RangeReader` implementations responsibilities as narrow as possible -- read the final ranges required for the requested tiles, but know nothing about tiles, then fill in the byte buffer at the appropriate places.  I'm thinking maybe this should be modified so that:

* The `CogTileInfo` object is passed to the range reader so that it has full knowledge of each requested tile index, offset, length, etc
* The range reader uses the `RangeBuilder` to build the contiguous ranges to be read
* The range reader then decomposes the results back into per-tile byte arrays (possibly storing them in the corresponding `TileRange` objects in `CogTileInfo`
* Update the `DefaultCogImageInputStream` accordingly with offset math, etc

What do you think about this solution?

Reply:
I think the RangeReader classes can be kept compact and focused, if instead of returning a byte[], end up returning a ``Map<long[], byte[]>`` with just the byte ranges read. The interesting and bit complicated bit will be updating the ``CogImageInputStream`` implementations so that they keep track of the current reading position and do the offset math dance. I don't see ``TileInfo`` needing to be updated, but it's more of an implementation preference.

Reply:
Updated to return `Map<Long, byte[]>`, where the Long value is the start of the range.  The end of the range, if needed, can be calculated using the byte[] length.

Reply:
Oh, I suppose it's important to mention that I completely removed a) any reference or attempt to understand the actual file size and b) usage of the class-level ByteBuffer to store all range data :)